### PR TITLE
Code refactoring

### LINKED ISSUE #1: [middle][Kodi Plugin] Code Refactor
The code needs to be refatored and for example the play part can be moved to an extra file! For any questions just ask me I will help as I can.

### PR DESCRIPTION
**Pull request by illright**
This PR improves code readability and maintainability by refactoring the addon functionality into a class.
Also it adds some advanced command-line args parsing for easier usage.

Fixes #1

### CODE REVIEW

File: addon.py [hunk 1]
```diff
@@ -1,78 +1,135 @@
-import sys
-import urllib
-import urlparse
+import argparse
+import os.path
+import urllib.parse
```
Comment (jugendhacker) [inline addon.py]: You need that urlparse because urllib.parse doesn't exist on a kodi, did you tested this?

File: addon.py [hunk 2]
```diff
@@ -1,78 +1,135 @@
...
+        description='Kodi add-on for mStream server communication'
+    )
+    parser.add_argument('base_url', help='launching URL of the add-on')
+    parser.add_argument('addon_handle', type=int, help='the handle of this add-on')
+    parser.add_argument('-m', '--mode',
+                        help='the operating mode of the add-on',
+                        choices=['files', 'play'])
+    parser.add_argument('-p', '--path',
+                        help='currently browsed path')
+    args = parser.parse_args()
```
Comment (jugendhacker) [inline addon.py]: The whole args thing doesn't work, I think because Kodi is calling the plugins in a diffrent way...

**Review by jugendhacker — CHANGES_REQUESTED** [review]
Please solve this problems, but I'm also not sure if you're familar with the programming of kodi plugins becuase this doesn't work at the moment! Please test it before you submit a pull request!

**Comment by illright** [comment]
I'm not familiar with writing Kodi plugins, just Python programming in general. Neither do I have a Kodi to test this..

The args thing can be fixed up to be called like the previous code.

About `urlparse`, which Python version is running on Kodi? Why is there no `urllib.parse`?

**Comment by jugendhacker** [comment]
You could easily setup a kodi for testing...

> The args thing can be fixed up to be called like the previous code.

Please change this so that this will work again

> About urlparse, which Python version is running on Kodi? Why is there no urllib.parse?

Not sure but I think their own Kind of python2

File: addon.py (hunk 2 above)
Comment (jugendhacker) [reply addon.py]: Still fails with
```
18:09:53.207 T:6296   ERROR: usage: [-h] base_url addon_handle {files,play} path
18:09:53.207 T:6296   ERROR: : error: argument addon_handle: invalid int value: ''
```
Please set you up a Kodi for testing, I can't do every testing for you, that's not the way how PRs work

**Comment by jugendhacker** [comment]
Doesn't work, please look at the comment, I provided a log

**Comment by jugendhacker** [comment]
Why you closed this PR?

**Comment by jugendhacker** [comment]
Your refactoring is good but we need to resolve the calling part...

**Comment by illright** [comment]
I do not own a Kodi so I simply can't test it. So we will have to compromise here.

**Comment by jugendhacker** [comment]
Kodi is a piece of software so you could install it on every pc...

**Comment by jugendhacker** [comment]
I really want to merge these changes but we need to rework the starup part before, you also don't need a kodi you could also read about how to do it [in the kodi wiki](https://kodi.wiki/view/Add-on_development)

**Comment by jugendhacker** [comment]
Please give me a answer....

**Comment by illright** [comment]
I'll get the argument parsing there way it was before, but I don't know
when I'll get around to that. Sorry for the delay

пн, 15 окт. 2018 г., 13:58 jugendhacker <notifications@github.com>:

**Comment by jugendhacker** [comment]
Ok no problem, I just thought you don't want to do this anymore...

**Comment by illright** [comment]
@jugendhacker should [this line](https://github.com/jugendhacker/plugin.audio.mstream/blob/master/addon.py#L78) contain `[0]`?

**Comment by jugendhacker** [comment]
> should this line contain [0]?

Yes because the path arg is an array with one entry... don't ask me why, I don't know it!

File: addon.py [hunk 3]
```diff
@@ -1,78 +1,130 @@
...
+
+if __name__ == '__main__':
+    print(sys.argv)
+
+    base_url = sys.argv[0]
+    addon_handle = int(sys.argv[1])
+    args = urlparse.parse_qs(sys.argv[2][1:])
+
+    operator = AddonOperator(addon_handle, base_url)
+    operator.act(args.get('mode'), args.get('path'))
```
Comment (jugendhacker) [inline addon.py]: There is a problem here when I call it the first time with mode=files because there is not path=...

**Review by jugendhacker — CHANGES_REQUESTED** [review]
Please get a better way for doing this... e.g giving the operator the whole args object...

**Comment by jugendhacker** [comment]
That seems good, but I have no time for testing now, will do it after 4pm UTC

**Comment by jugendhacker** [comment]
Still getting
```
Error Type: <type 'exceptions.TypeError'>
Error Contents: 'NoneType' object has no attribute '__getitem__'
 Traceback (most recent call last):
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 130, in <module>
   operator.act(args.get('mode', None), args.get('path', None))
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 119, in act
   self.play_song(path[0])
 TypeError: 'NoneType' object has no attribute '__getitem__'
```

**Comment by jugendhacker** [comment]
Next error incomming
```
Error Type: <type 'exceptions.KeyError'>                                                                            
Error Contents: ('path',)                                                                                           
Traceback (most recent call last):                                                                                  
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 130, in <module>           
    operator.act(args.get('mode', None), args.get('path', None))                                                    
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 113, in act                
    self.list_folder_contents(self.folders[0])                                                                      
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 66, in list_folder_contents
    path=os.path.join(file['path'], file['name'])                                                                   
KeyError: ('path',)                                                                                                 
```

**Comment by jugendhacker** [comment]
```
Error Type: <type 'exceptions.TypeError'>                                                                           
Error Contents: list indices must be integers, not str                                                              
Traceback (most recent call last):                                                                                  
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 132, in <module>           
    operator.act(args.get('mode', None), args.get('path', None))                                                    
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 115, in act                
    self.list_folder_contents(self.folders[0])                                                                      
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 66, in list_folder_contents
    path=os.path.join(response['contents']['path'],                                                                 
TypeError: list indices must be integers, not str                                                                   
```
New Error xD

**Comment by jugendhacker** [comment]
Sorry was offline by a technical issue with my Laptop, will test your latest commit tomorrow!

**Comment by jugendhacker** [comment]
Now getting
```
Error Type: <type 'exceptions.UnicodeEncodeError'>                                                                  
Error Contents: 'ascii' codec can't encode character u'\u2019' in position 12: ordinal not in range(128)            
Traceback (most recent call last):                                                                                  
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 127, in <module>           
    operator.act(args.get('mode', None), args.get('path', None))                                                    
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 110, in act                
    self.list_folder_contents(self.folders[0])                                                                      
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 65, in list_folder_contents
    path=response['path'] + file['name']                                                                            
  File "C:\Users\Julian\AppData\Roaming\Kodi\addons\plugin.audio.mstream\addon.py", line 19, in build_url           
    return self.base_url + '?' + urllib.urlencode(query)                                                            
  File "C:\Program Files (x86)\Kodi\system\python\Lib\urllib.py", line 1343, in urlencode                           
    v = quote_plus(str(v))                                                                                          
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2019' in position 12: ordinal not in range(128)        
```

**Comment by jugendhacker** [comment]
Now it does something wrong with the url Encoding...
It requests `http://server:3000/media/music%2FAdele%2FSkyfall.mp3?token=...` instead of `http://server:3000/media/music/Adele/Skyfall.mp3?token=...`

**Comment by illright** [comment]
The good intentions are at fault :p shouldn't have changed any functions

**Comment by jugendhacker** [comment]
The Problem is this strange kodi python engine xD

**Comment by jugendhacker** [comment]
But thank you, now everything works!

**Review by jugendhacker — APPROVED** [review]

**Pull request merged**